### PR TITLE
Toggable fields

### PR DESCRIPTION
**Pull request by iMokhles**
So i had this in mind for months but didn't have much time to implement it **SORRY**

so the goal here is to make ( **ALL** ) fields toggable without extra field 

you can use ( **ANY** ) field value as a condition value to toggle another field visibility 

# **The problem and solution**

let's say we have a price field ( called ) price before as i added in the example below a hint under the field say's that leave it empty if it's a free package but the field is really became not useful now as it's shown for no propose ( if the user decided to leave it blank ) so why not hiding it from the form and also disabling it from sending it's value with the ( **SAVE** ) request ? it was complicated to acheive such concept before but now it's super easy all you have to do is attaching the child field to another field's value to make sure the child field is only visible if the ( **PARENT** ) field's value is equal to what you have set, so how you can do that, **EASY** add visibility option to your ( **CHILD** ) field

#### Here we add the visibility to the CHILD and attach it to another field ( BY ) name then we telling the conditional value which the ( **PARENT** )  field should be setted to to make the ( **CHILD** ) field visible
```
"visibility" => [
                'field_name' => 'is_free',
                'value'      => 'paid',
                'add_disabled' => true, // if you need to disable this field value to not be send through create/update request set it to true, otherwise set it to true 

            ],
```

#### Here's a ( **CHILD** ) field example
```
CRUD::addField([   // Number
            'name' => 'price',
            'label' => 'Price',
            'type' => 'number',

            // optionals
            'attributes' => ["step" => "any"], // allow decimals
            'prefix'     => "$",
            'hint'      => 'leave empty if it\'s a Free package',
            'suffix'     => "U.S Dollar",
            "visibility" => [
                'field_name' => 'is_free',
                'value'      => 'paid',
                'add_disabled' => true, // if you need to disable this field value to not be send through create/update request set it to true, otherwise set it to true 
            ],
            'value' => 1.99,
            'tab' => 'Status',
        ]);
```

#### Here's a ( **PARENT** ) field example as select_from_array
```
CRUD::addField([   // select_from_array
            'name'        => 'is_free',
            'label'       => "Is Free",
            'type'        => 'select_from_array',
            'options'     => ['free' => 'Free', 'paid' => 'Paid'],
            'allows_null' => false,
            'default'     => 'free',
            'tab' => 'Status',
        ]);
```

One the selected field changed to paid the price field will be visible with a toggle animation ( thanks JQuery )
here's a video of it as Normal field

https://user-images.githubusercontent.com/1247248/103076199-78122780-45cd-11eb-8510-3278216239c2.mov

And here's a video and proof that it works inside Repeatable field too

https://user-images.githubusercontent.com/1247248/103076268-9f68f480-45cd-11eb-9b99-13048f271d03.mov

Here's an example shows the use case as product's multiple variants toggling ;)

https://user-images.githubusercontent.com/1247248/103181099-74b4bf80-489d-11eb-904b-f2e735311392.mov

Checklist example

https://user-images.githubusercontent.com/1247248/103183470-febb5300-48b2-11eb-9444-9a0e5e01c95b.mov

code
```
[   // Checklist
                'label'     => trans('backpack::permissionmanager.roles'),
                'type'      => 'checklist',
                'name'      => 'roles',
                'entity'    => 'roles',
                'attribute' => 'name',
                'model'     => config('permission.models.role'), // foreign key model
                'pivot'     => true,
            ],
            [   // Checklist
                'label'     => ucfirst(trans('backpack::permissionmanager.permission_singular')),
                'type'      => 'checklist',
                'name'      => 'permissions',
                'entity'    => 'permissions',
                'attribute' => 'name',
                'model'     => config('permission.models.permission'), // foreign key model
                'pivot'     => true,
                "visibility" => [
                    'field_name' => 'roles',
                    'value'      => 4,
                    'add_disabled' => true,
                ],
            ],
```

i didn't work that much and i didn't try to fix issues ( Hope there are no issues ) 
such functionality shouldn't break anything ( i hope so )

**I KNOW** that @tabacitu asked this to be an addon but creating an addon for such functionality is really hard to do ( What do you think ) @tabacitu  ?

DIDN'T test it with Inline Create ( but from what i see with Repeatable field and normal field it should work without issue !

Note:
child fields inside Repeatable don't accept attr() from Jquery to add and remove disabled attribute ( didn't try any solution tell now ) as i'm not needing it inside repeatable at all ( **FOR NOW** )

### CODE REVIEW

**Comment by soufiene-slimi** [comment]
Hi @iMokhles, thanks for this [pull](https://github.com/Laravel-Backpack/CRUD/pull/3420).

This really looks great, in many scenarios, you will definitely need this feature.
Just wondering if we can improve the functionality, maybe I want to show the price input if the user accepts the rules of posting a new product (a checkbox field) and choose to create a paid product, so this scenario makes me think and ask the following questions:
- Is having multiple parent fields values will be better? and if so, will it be difficult to achieve instead of the single-parent value condition? 
- Is there a scenario where I will need to do another check before showing the price field (maybe some calculation or some database query or a specific date)

And if I consider that those are easy to implement, I think that version 4.2 will be a perfect place to introduce this feature.
So what do you think @tabacitu ?

**Comment by iMokhles** [comment]
Here's an example shows the use case as product's multiple variants toggling ;)

https://user-images.githubusercontent.com/1247248/103181099-74b4bf80-489d-11eb-904b-f2e735311392.mov

**Comment by iMokhles** [comment]
Checklist example

https://user-images.githubusercontent.com/1247248/103183470-febb5300-48b2-11eb-9444-9a0e5e01c95b.mov

code
```
[   // Checklist
                'label'     => trans('backpack::permissionmanager.roles'),
                'type'      => 'checklist',
                'name'      => 'roles',
                'entity'    => 'roles',
                'attribute' => 'name',
                'model'     => config('permission.models.role'), // foreign key model
                'pivot'     => true,
            ],
            [   // Checklist
                'label'     => ucfirst(trans('backpack::permissionmanager.permission_singular')),
                'type'      => 'checklist',
                'name'      => 'permissions',
                'entity'    => 'permissions',
                'attribute' => 'name',
                'model'     => config('permission.models.permission'), // foreign key model
                'pivot'     => true,
                "visibility" => [
                    'field_name' => 'roles',
                    'value'      => 4,
                    'add_disabled' => true,
                ],
            ],
```

**Comment by promatik** [comment]
Hi @iMokhles sorry for taking so long to answer, the toggable fields is a really cool feature that I've had to implement my self several times, I'm looking forward to merge this 👏

Right now we're on a sprint to finish 4.1, and I'll certainly take a look at this real soon on v4.2

**Comment by zachweix** [comment]
On my first test it failed because I have a date range, which meant that in the wrapper start file, we are trying to set the id as an array

**Comment by iMokhles** [comment]
any update ?

**Comment by resulparahadow** [comment]
Hello everyone!
It didn't work in my project. I put it in updateOperation. It doesn't make sense
`    protected function setupCreateOperation()
    {
        CRUD::addField([   // Number
            'name' => 'price',
            'label' => 'Price',
            'type' => 'number',

            // optionals
            'attributes' => ["step" => "any"], // allow decimals
            'prefix'     => "$",
            'hint'      => 'leave empty if it\'s a Free package',
            'suffix'     => "U.S Dollar",
            "visibility" => [
                'field_name' => 'is_free',
                'value'      => 'paid',
                'add_disabled' => true, // if you need to disable this field value to not be send through create/update request set it to true, otherwise set it to true
            ],
            'value' => 1.99,
            'tab' => 'Status',
        ]);
        CRUD::addField([   // select_from_array
            'name'        => 'is_free',
            'label'       => "Is Free",
            'type'        => 'select_from_array',
            'options'     => ['free' => 'Free', 'paid' => 'Paid'],
            'allows_null' => false,
            'default'     => 'free',
            'tab' => 'Status',
        ]);
    }`

**Review by resulparahadow — APPROVED** [review]

**Review by resulparahadow — APPROVED** [review]

**Comment by iMokhles** [comment]
@tabacitu any updates on this !!!

**Comment by pxpm** [comment]
@iMokhles hey man. 

I am not @tabacitu sorry! But me and @tabacitu had a great conversation about this at the start of the week. 

This feature had so much more to say than what it shows here. Hehe

We haven't reached a final solution, but we wondered about the fact that `toggable fields` is just one part of the `interactive fields`. What I mean is, apart from `show/hide (toggable)`: 
- people probably would want to show/hide more than one field.
- people would like to have the field border red or something (not show/hide)
- people would like to calculate a total for some field when some other change 
- people would like to mark the checkbox instead of show/hide it

We have some scenarios on the table, I will just quickly describe to you one that I suggested to @tabacitu 

I would create 3/4 predefined javascript functions:
- `showField` / `hideField`
-  `setFieldValue` / `setAttribute` (this could replace the setValue and treat everything as an attribute, including value.

Everytime a field is "changed/clicked", a request is made to the backend where developer configured the fields:

```php
$this->crud->addField([
    'name' => 'some_field',
    'interaction' => function($changes) {
        if($changes['checkbox']) {
            $this->addJsFunction('showField', 'field_name');
            $this->addJsFunction('setAttribute', 'attribute', 'value);
        }
    }])
```
So basically this is just `Livewire` but way worse than `Livewire`. So we are still trying to figure out the best solution, I just shared this one with you so you know where we are trying to go. If we should settle only on `show/hide` fields etc etc. 

Thanks for sticking with us @iMokhles , really appreciated!

Pedro

**Comment by ASVVIZIT** [comment]
Hello gentlemen. I would like to know if work is being done on this functionality and is there a more detailed use case for version 4.2. And how soon will laravel backpak 4.2 be released? Will this functionality be implemented in the core of this version.
In our application, a huge number of fields with conditions. Show/hide select as usual, show/hide group of fields as multiple choice. Visual formation is also of the type of Tabs, for example, as Tabs in a vertical arrangement inside Tab level 1 (I hope I clearly wrote what I meant). Checkboxes are also needed.

**Comment by tabacitu** [comment]
@iMokhles I'm super-greteful for everything you've done here. I think your solution covers more ground than ANY of the previous solutions... but I think it would still fall short from complete customizability. Right after launch, I expect people to come up and say "_yes but... I need it to do that bit differently_". Which is the reason I've refrained from doing anything about it so far, in the Backpack core... But since people voted on it... let's do this! What do you think of my alternative in https://github.com/Laravel-Backpack/CRUD/issues/4158 ?

**Comment by tabacitu** [comment]
@iMokhles I've submitted the prototype for this feature in https://github.com/Laravel-Backpack/CRUD/pull/4312 - it's a simple JS library that will allow us devs to interact with fields in JS (not PHP). I think that's a first step towards building what you did here, but in a more general way, that covers even more ground than yours did. So that it applies to all fields, and helps in all use cases, even the most complex. Basically, let's do the JS syntax first, then the PHP syntax afterwards. That's what I'm saying 😀

**If you have time, I'd appreciate your feedback on that PR.** I really want to get it right. After that's merged, we'll think about a PHP syntax to do the same, and your experience would also be useful there.

Let me know what you think! I'll close this PR in favor of that one.
Cheers! And of course... thank you for the effort and patience 🙏
